components/script/dom/webglrenderingcontext.rs

```diff
@@ -644,6 +643,57 @@ impl WebGLRenderingContext {
         }
     }
 
+    // Remove premultiplied alpha.
+    // This is only called when texImage2D is called using a canvas2d source and
+    // UNPACK_PREMULTIPLY_ALPHA_WEBGL is disabled. Pixels got from a canvas2D source
+    // are always RGBA8 with premultiplied alpha, so we don't have to worry about
+    // additional formats as happens in the premultiply_pixels method.
+    fn remove_premultiplied_alpha(&self, mut pixels: Vec<u8>) -> Vec<u8> {
+        for rgba in pixels.chunks_mut(4) {
+            let a = (rgba[3] as f32) / 255.0;
+            rgba[0] = (rgba[0] as f32 / a) as u8;
+            rgba[1] = (rgba[1] as f32 / a) as u8;
+            rgba[2] = (rgba[2] as f32 / a) as u8;
+        }
+        pixels
+    }
+
+    fn prepare_pixels(&self,
+                      internal_format: TexFormat,
+                      data_type: TexDataType,
+                      width: u32,
+                      height: u32,
+                      unpacking_alignment: u32,
+                      source_premultiplied: bool,
+                      mut pixels: Vec<u8>) -> Vec<u8> {
+        let dest_premultiply = self.texture_unpacking_settings.get().contains(PREMULTIPLY_ALPHA);
+        if !source_premultiplied && dest_premultiply {
+            pixels = self.premultiply_pixels(internal_format, data_type, pixels);
+        }
+        else if source_premultiplied && !dest_premultiply {
```

Comment:
I mean, the else should be after the brace, I'm not suggesting to change the condition.